src/main/java/de/chojo/jdautil/pagination/PageService.java

```diff
@@ -0,0 +1,230 @@
+/*
+ *     SPDX-License-Identifier: AGPL-3.0-only
+ *
+ *     Copyright (C) 2022 RainbowDashLabs and Contributor
+ */
+
+package de.chojo.jdautil.pagination;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import de.chojo.jdautil.localization.ILocalizer;
+import de.chojo.jdautil.pagination.bag.IPageBag;
+import de.chojo.jdautil.pagination.exceptions.EmptyPageBagException;
+import net.dv8tion.jda.api.entities.ChannelType;
+import net.dv8tion.jda.api.entities.Emoji;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.GuildMessageChannel;
+import net.dv8tion.jda.api.entities.Message;
+import net.dv8tion.jda.api.entities.MessageChannel;
+import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+import net.dv8tion.jda.api.interactions.InteractionHook;
+import net.dv8tion.jda.api.interactions.callbacks.IReplyCallback;
+import net.dv8tion.jda.api.interactions.components.ActionRow;
+import net.dv8tion.jda.api.interactions.components.buttons.Button;
+import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle;
+import net.dv8tion.jda.api.sharding.ShardManager;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+public class PageService extends ListenerAdapter {
+    private final String previousLabel;
+    private final String nextLabel;
+    private final String previousText;
+    private final String nextText;
+    private final Cache<Long, IPageBag> cache;
+    private final ILocalizer localizer;
+
+    public PageService(String previousLabel, String nextLabel, String previousText, String nextText, Cache<Long, IPageBag> cache, ILocalizer localizer) {
```

Comment:
Should this constructor be private because you use a builder below?

Reply:
Yes it should